src/test/java/org/mastodon/mamut/treesimilarity/NodeMappingTest.java

```diff
@@ -0,0 +1,154 @@
+package org.mastodon.mamut.treesimilarity;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.Test;
+import org.mastodon.mamut.treesimilarity.tree.SimpleTree;
+import org.mastodon.mamut.treesimilarity.tree.SimpleTreeExamples;
+import org.mastodon.mamut.treesimilarity.tree.Tree;
+import org.mastodon.mamut.treesimilarity.tree.TreeUtils;
+
+public class NodeMappingTest
+{
+
+	private static final BiFunction< Double, Double, Double > DEFAULT_COSTS = ( o1, o2 ) -> {
+		if ( o2 == null )
+			return o1;
+		else
+			return Math.abs( o1 - o2 );
+	};
+
+	@Test
+	public void testTree1vs2()
+	{
+		testNodeMapping( SimpleTreeExamples.tree1(), SimpleTreeExamples.tree2(),
+				20, "10->10, 20->30, 30->20" );
+	}
+
+	@Test
+	public void testTree3vs4()
+	{
+		testNodeMapping( SimpleTreeExamples.tree3(), SimpleTreeExamples.tree4(),
+				4, "1->1, 1->1, 100->100" );
+	}
+
+	@Test
+	public void testTree5vs6()
+	{
+		testNodeMapping( SimpleTreeExamples.tree5(), SimpleTreeExamples.tree6(),
+				49, "13->12, 203->227, 203->227" );
+	}
+
+	@Test
+	public void testTree5vs7()
+	{
+		testNodeMapping( SimpleTreeExamples.tree5(), SimpleTreeExamples.tree7(),
+				69, "13->12, 203->227, 203->227" );
+	}
+
+	@Test
+	public void testTree8vs9()
+	{
+		testNodeMapping( SimpleTreeExamples.tree8(), SimpleTreeExamples.tree9(),
+				1, "1->1, 2->2, 3->3, 4->4, 5->4, 8->8, 8->8" );
+	}
+
+	@Test
+	public void testTree12vs13()
+	{
+		testNodeMapping( SimpleTreeExamples.tree12(), SimpleTreeExamples.tree13(),
+				2, "100->100, 1000->1000, 200->200" );
+	}
+
+	@Test
+	public void testNonBinaryTree()
+	{
+		testNodeMapping( SimpleTreeExamples.tree14(), SimpleTreeExamples.nonBinaryTree(),
+				1_000_003, "10000->10001, 2->3, 4->5" );
+	}
+
+	@Test
+	public void testTree15vs16()
```

Comment:
I am wondering, why this test shows that one should expect only 3 mappings, while there are 5 nodes in each tree.

![grafik](https://github.com/mastodon-sc/mastodon-deep-lineage/assets/10515534/046a59c0-5433-41d5-9229-b22a040755e6)

I assume there is a simple explanation and I would be interested in it.

Reply:
The zhang tree edit distance only allows _restricted_ node mappings. Here is the relevant definition in the paper:

![image](https://github.com/mastodon-sc/mastodon-deep-lineage/assets/24407711/b02d96eb-402b-4cd1-b37c-fd96714999b0)

Here is my translation. The common parent of two trees must be mapped to a common parent of the to mapped trees. Something like that. I don't yet understand these details either.

These rules forbid to map for example 1000 -> 1000, 100 -> 100, 200 -> 200, 300 -> 300. So the mapping that maps the 3 nodes is the mapping with the lowest cost that is allowed.